src/platform/backends/lxd/lxd_virtual_machine.cpp

```diff
@@ -0,0 +1,386 @@
+                                   {"protocol", "simplestreams"},
+                                   {"fingerprint", QString::fromStdString(desc.image.id)}}}};
+
+        auto json_reply = lxd_request(manager, "POST", QUrl(QString("%1/virtual-machines").arg(base_url.toString())),
+                                      virtual_machine);
+        mpl::log(mpl::Level::debug, name.toStdString(),
+                 fmt::format("Got LXD creation reply: {}", QJsonDocument(json_reply).toJson()));
+
+        if (json_reply["metadata"].toObject()["class"] == QStringLiteral("task") &&
+            json_reply["status_code"].toInt(-1) == 100)
+        {
+            QUrl task_url(QString("%1/operations/%2")
+                              .arg(base_url.toString())
+                              .arg(json_reply["metadata"].toObject()["id"].toString()));
+
+            // Instead of polling, need to use websockets to get events
+            while (true)
+            {
+                try
+                {
+                    auto task_reply = lxd_request(manager, "GET", task_url);
+
+                    if (task_reply["error_code"].toInt(-1) != 0)
+                    {
+                        break;
+                    }
+
+                    auto status_code = task_reply["metadata"].toObject()["status_code"].toInt(-1);
+                    if (status_code == 200)
+                    {
+                        break;
+                    }
+                    else
+                    {
+                        std::this_thread::sleep_for(5s);
+                    }
+                }
+                // Implies the task is finished
+                catch (const LXDNotFoundException& e)
+                {
+                    break;
+                }
+            }
+        }
+
+        current_state();
+    }
+}
+
+mp::LXDVirtualMachine::~LXDVirtualMachine()
+{
+    update_suspend_status = false;
+
+    if (current_state() == State::running)
+        suspend();
+}
+
+void mp::LXDVirtualMachine::start()
+{
+    auto present_state = current_state();
+
+    if (present_state == State::running)
+        return;
+
+    request_state("start");
+
+    state = State::starting;
+    update_state();
+}
+
+void mp::LXDVirtualMachine::stop()
+{
+    std::unique_lock<decltype(state_mutex)> lock{state_mutex};
+    auto present_state = current_state();
+
+    if (present_state == State::running || present_state == State::delayed_shutdown)
+    {
+        auto state_task = request_state("stop");
+        if (state_task["metadata"].toObject()["class"] == QStringLiteral("task") &&
+            state_task["status_code"].toInt(-1) == 100)
+        {
+            QUrl task_url(QString("%1/operations/%2/wait")
+                              .arg(base_url.toString())
+                              .arg(state_task["metadata"].toObject()["id"].toString()));
+            lxd_request(manager, "GET", task_url);
```

Comment:
Couldn't use that instead of the above poll? Or will it time out?

Reply:
I'm not sure if you're referring to using `wait` here or the polling method.

Reply:
Yes, using `wait`, why not use it above?

Reply:
Hmm, right, I think using a `wait` with no or a long-ish timeout above would be just as effective.  I observed that LXD can take quite a long time uncompressing an image for the first time.